cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java

```diff
@@ -449,37 +441,160 @@ public void onSuccess(V1Status v1Status, int statusCode, Map<String, List<String
+    CompletableFuture<KubeTwillController> resultFuture;
+    CompletionStage<String> completionStage;
+    resultFuture = new CompletableFuture<>();
+    // delete the resource
+    completionStage = deleteResource();
+    completionStage.whenComplete((name, t) -> {
+      if (t != null) {
+        resultFuture.completeExceptionally(t);
+      } else {
+        resultFuture.complete(KubeTwillController.this);
+      }
+    });
+    // If the deletion was successful, reflect the status into the completion future
+    // If the deletion was failed, don't change the completion state so that caller can retry
+    resultFuture.thenAccept(completion::complete);
+    return resultFuture;
+  }
+
+  /**
+   * Cleans up the job and returns future with correct job status.
+   */
+  private CompletableFuture<KubeTwillController> cleanupJob() {
+    // get job status
+    CompletionStage<String> completionStage = getJobCompletionStatus();
+    completionStage.whenComplete((name, t) -> {
+      if (t != null) {
+        completion.completeExceptionally(t);
+      } else {
+        completion.complete(KubeTwillController.this);
+      }
+    });
+    return completion;
+  }
+
+  /**
+   * Returns job completion status.
+   */
+  private CompletionStage<String> getJobCompletionStatus() {
+    CompletableFuture<String> resultFuture = new CompletableFuture<>();
+    String name = meta.getName();
+    // In case we fail to get job status due to some transient issue, retry 5 times. There is no need to retry for too
+    // long otherwise corresponding program may appear to be stuck while retrying to get job status.
+    // If we exhaust retries, complete future exceptionally. Though in normal scenario, retries is not expected
+    // to be performed.
+    int retryCount = 5;
+    do {
+      try {
+        V1Job v1Job = batchV1Api.readNamespacedJob(name, kubeNamespace, null, null, null);
+        V1JobStatus status = v1Job.getStatus();
+
+        // If job has failed, mark future as failed. Else mark it as succeeded.
+        if (status != null && status.getFailed() != null) {
+          resultFuture.completeExceptionally(new RuntimeException(String.format("Job %s has failed status.", name)));
+        } else {
+          resultFuture.complete(name);
+        }
+        retryCount = 0;
+
+        // Also attempt to delete the job once status is available
+        deleteJob(meta).whenComplete((jName, t) -> {
+          if (t != null) {
+            LOG.warn("Failed to delete job {}. Attempt will be retried", jName);
+          } else {
+            LOG.trace("Successfully deleted job {}", jName);
+          }
+        });
+      } catch (ApiException e) {
+        if (e.getCode() == 404) {
+          LOG.warn("Failed to get job status because job {} was not found.", name, e);
+          // There is no need to retry if job itself is not present. Mark the job as failed.
+          retryCount = 0;
+          completeExceptionally(resultFuture, e);
```

Comment:
Do we still want to complete with exception?

Reply:
This could happen in 2 cases:
1. user manually deletes the job before controller completes successfully
2. job cleaner deletes the job before controller completes. 

I thought in either case, if we dont know the job status, we mark job as failed. 

To avoid this race, we can rely on completion time of the job. Anything before time t should be deleted by job cleaner. That way we can be certain that this would only happen for case #1

Reply:
Had offline discussion with @chtyim and @albertshau 
In case of 404 (job deletion/status) we always mark the pipeline as KILLED and hence stopped status on UI.

![image](https://user-images.githubusercontent.com/14131070/126718774-75d6d8e7-c2ab-48fe-bf29-5dfc3fcf3004.png)

![image](https://user-images.githubusercontent.com/14131070/126718785-2d02fce3-8215-4a74-a1bc-b7a59fcb2eee.png)